proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,664 @@
+
+An async currently always executes on the global concurrent executor, and with the inclusion of this proposal, it does take into account task executor preference. In other words, if an executor preference is set, it will be used by async let to enqueue its underlying task:
+
+```swift
+func test(_ someExecutor: any Executor) -> Int {
+  someExecutor.assertIsolated()
+  return 42
+}
+
+await withTaskExecutor(someExecutor) { 
+  async let value = test(someExecutor) // executes on 'someExecutor'
+  // ... 
+  await value
+}
+```
+
+### Task executor preference and TaskGroups
+
+A `TaskGroup` and its various friends (`ThrowingTaskGroup`, `DiscardingTaskGroup`, ...) are the most powerful, but also most explicit and verbose API for structured concurrency. A group allows creating multiple child tasks using the `addTask` method, and always awaits all child tasks to complete before returning.
+
+This proposal adds overloads to the `addTask` method, which changes the executor the child tasks will be enqueued on:
+
+```swift
+extension (Discarding)(Throwing)TaskGroup { 
+  mutating func addTask(
+    on executor:  (any Executor)?,  // 'nil' means the global pool
+    priority:  TaskPriority? = nil,
+  operation:  @Sendable @escaping () async (throws) -> Void
+  )
+  
+  mutating func addTask<TargetActor>(
+ on actor: TargetActor,
+    priority: TaskPriority? = nil,
+    operation: @Sendable @escaping (isolated TargetActor) async (throws) -> Void
+  ) where TargetActor: Actor
+}
+```
+
+Which allows users to require child tasks be enqueued and run on specific executors:
+
+```swift
+Task(on: specialExecutor) {
+  _ = await withTaskGroup(of: Int.self) { group in 
+    group.addTask() {
+      specialExecutor.assertIsolated()
+      12 
+    }
+    group.addTask(on: differentExecutor) {
+      differentExecutor.assertIsolated()
+      42 
+    }
+    group.addTask(on: nil) { 
+      // guaranteed to run on the default global concurrent executor;
+      // this is equivalent to the only behavior a group exhibits today.
+      84
+    } 
+    return await group.next()!
+  }
+```
+
+This gives developers explicit control over where a task group child task shall be executed. Notably, this gives callers of libraries more control over where work should be performed. Do note that code executing on an actor will always hop to that actor; and task executor preference has no impact on code which *requires* to be running in some specific isolation.
+
+If a library really wants to ensure that hops to the global concurrent executor *are* made by e.g. such task group, they should use `group.addTask(on: nil)` to override the inherited task executor preference.
+
+### Task executor preference and Unstructured Tasks
+
+We propose adding new APIs and necessary runtime changes to allow a Task to be enqueued directly on a specific `Executor`, by using a new `Task(on:)` initializer:
+
+```swift
+extension Task where Failure == Never {
+  @discardableResult
```

Comment:
We have to see how we fold this in with the typed throws proposal since we are dropping a few inits and the `@discarableResult` there.

Reply:
Yeah I'm not including any of those changes but yes they will have to mesh together